data/src/main/scala/ch.epfl.scala.index.data/download/PlayWsDownloader.scala

```diff
@@ -107,10 +94,69 @@ trait PlayWsDownloader {
       )
     }
 
+    downloadItems(toDownload, parallelism, message, processItem)
+  }
+
+  /**
+    * Actual download of bunch of documents from Github's GraphQL API. Will loop through all and display a status bar in the console output.
+    *
+    * @param message the message for the loader info
+    * @param toDownload the set of downloadable elements
+    * @param downloadUrl a function to get the WsRequest for the current element
+    * @param query query sent to Github's GraphQL API
+    * @param process a function to process the response in succes case
+    * @tparam T Input type
+    * @tparam R output type
+    */
+  @throws(classOf[Exception])
+  def downloadGraphql[T, R](
+      message: String,
+      toDownload: Set[T],
+      downloadUrl: AhcWSClient => WSRequest,
+      query: T => JsObject,
+      process: (T, WSResponse) => Try[R],
+      parallelism: Int
+  ): Seq[R] = {
+
+    def processItem(item: T, client: AhcWSClient, progressBar: ProgressBar) = {
+      val request = downloadUrl(client)
+      val response = request.post(query(item))
+      response.transform(
+        data => {
+          if (toDownload.size > 1) {
+            progressBar.step()
+          }
+          process(item, data) match {
+            case Success(result) => result
+            case Failure(ex) => throw ex
```

Comment:
True. The whole reason I'm throwing the exception in the first place is so that when a thread [hits](https://github.com/scalacenter/scaladex/pull/432/files#diff-e03c541cf1bd7ec0322a9a6571160bebR325) the API rate limit, it will throw the error which will get caught by the parent thread in `PlayWsDownloader` and will then stop all the other threads from executing (since when the rate limit has been reached, those other threads will get errors when they hit the API so no point in them executing anymore).

Is there a better way to do this @julienrf? Maybe sending messages between actors ... I tried looking into that but couldn't figure out anything. Or we could pause all the threads until the API rate limit resets (maximum time of 1 hour).